Author: Krithi D Shetty <kdshetty@qti.qualcomm.com>
Date:   Tue Oct 21 15:02:29 2025 +0530

    [iproute2] Add suport to display EHT database

    "-e" flag for "bridge mdb show" command to display EHT information including host IP address, host MAC address, mode, and source list.

    Change-Id: Id04ec124d09ed38a41fc17ec0ab53c8b50340195
    Signed-off-by: KRITHI D SHETTY <kdshetty@qti.qualcomm.com>
---
Index: iproute2-6.11.0/bridge/br_common.h
===================================================================
--- iproute2-6.11.0.orig/bridge/br_common.h
+++ iproute2-6.11.0/bridge/br_common.h
@@ -32,3 +32,4 @@ extern int timestamp;
 extern int compress_vlans;
 extern int json;
 extern struct rtnl_handle rth;
+extern int show_eht_hosts;
Index: iproute2-6.11.0/bridge/bridge.c
===================================================================
--- iproute2-6.11.0.orig/bridge/bridge.c
+++ iproute2-6.11.0/bridge/bridge.c
@@ -28,6 +28,7 @@ int json;
 int timestamp;
 static const char *batch_file;
 int force;
+int show_eht_hosts;
 
 static void usage(void) __attribute__((noreturn));
 
@@ -39,7 +40,8 @@ static void usage(void)
 "where  OBJECT := { link | fdb | mdb | mst | vlan | vni | monitor }\n"
 "       OPTIONS := { -V[ersion] | -s[tatistics] | -d[etails] |\n"
 "                    -o[neline] | -t[imestamp] | -n[etns] name |\n"
-"                    -com[pressvlans] -c[olor] -p[retty] -j[son] }\n");
+"                    -com[pressvlans] -c[olor] -p[retty] -j[son] |\n"
+"                    -e (show EHT hosts in mdb show) }\n");
 	exit(-1);
 }
 
@@ -167,6 +169,8 @@ main(int argc, char **argv)
 			if (argc <= 1)
 				usage();
 			batch_file = argv[1];
+		} else if (matches(opt, "-e") == 0) {
+			++show_eht_hosts;
 		} else {
 			fprintf(stderr,
 				"Option \"%s\" is unknown, try \"bridge help\".\n",
Index: iproute2-6.11.0/bridge/mdb.c
===================================================================
--- iproute2-6.11.0.orig/bridge/mdb.c
+++ iproute2-6.11.0/bridge/mdb.c
@@ -36,7 +36,7 @@ static void usage(void)
 		"Usage: bridge mdb { add | del | replace } dev DEV port PORT grp GROUP [src SOURCE] [permanent | temp] [vid VID]\n"
 		"              [ filter_mode { include | exclude } ] [ source_list SOURCE_LIST ] [ proto PROTO ] [ dst IPADDR ]\n"
 		"              [ dst_port DST_PORT ] [ vni VNI ] [ src_vni SRC_VNI ] [ via DEV ]\n"
-		"       bridge mdb {show} [ dev DEV ] [ vid VID ]\n"
+		"       bridge {-e} mdb {show} [ dev DEV ] [ vid VID ]\n"
 		"       bridge mdb get dev DEV grp GROUP [ src SOURCE ] [ vid VID ] [ src_vni SRC_VNI ]\n"
 		"       bridge mdb flush dev DEV [ port PORT ] [ vid VID ] [ src_vni SRC_VNI ] [ proto PROTO ]\n"
 		"              [ [no]permanent ] [ dst IPADDR ] [ dst_port DST_PORT ] [ vni VNI ]\n");
@@ -152,6 +152,127 @@ static void print_src_entry(struct rtatt
 	close_json_object();
 }
 
+static void print_eht_host_src_entry(struct rtattr *src_attr, int af)
+{
+	struct rtattr *stb[MDBA_MDB_EATTR_EHT_HOST_SRC_MAX + 1];
+	SPRINT_BUF(abuf);
+	const char *addr;
+
+	parse_rtattr_nested(stb, MDBA_MDB_EATTR_EHT_HOST_SRC_MAX, src_attr);
+	if (!stb[MDBA_MDB_EATTR_EHT_HOST_SRC_ADDR])
+		return;
+
+	addr = inet_ntop(af, RTA_DATA(stb[MDBA_MDB_EATTR_EHT_HOST_SRC_ADDR]),
+			abuf, sizeof(abuf));
+	if (!addr)
+		return;
+
+	open_json_object(NULL);
+	print_color_string(PRINT_ANY, ifa_family_color(af),
+			"source_address", "%s\t", addr);
+
+	close_json_object();
+}
+
+static void print_eht_host_srcs(struct rtattr *srcs_attr, int af)
+{
+	struct rtattr *i;
+	int rem;
+
+	if (!srcs_attr)
+		return;
+
+	open_json_array(PRINT_JSON, "sources");
+	if (!is_json_context())
+		print_string(PRINT_FP, NULL, "%s", "      Sources:\t");
+
+	rem = RTA_PAYLOAD(srcs_attr);
+	for (i = RTA_DATA(srcs_attr); RTA_OK(i, rem); i = RTA_NEXT(i, rem)) {
+		print_eht_host_src_entry(i, af);
+	}
+
+	print_nl();
+	close_json_array(PRINT_JSON, NULL);
+}
+
+static void print_eht_host_entry(struct rtattr *host_attr, int af)
+{
+	struct rtattr *htb[MDBA_MDB_EATTR_EHT_HOST_MAX + 1];
+	SPRINT_BUF(ip_buf);
+	SPRINT_BUF(mac_buf);
+	const char *ip_addr = NULL;
+	const char *mac_addr = NULL;
+	const char *mode_str;
+	__u8 mode;
+	__u32 nsrcs;
+
+	parse_rtattr_nested(htb, MDBA_MDB_EATTR_EHT_HOST_MAX, host_attr);
+
+	open_json_object(NULL);
+
+	/* Parse and print host IP address */
+	if (htb[MDBA_MDB_EATTR_EHT_HOST_IP_ADDR]) {
+		const void *ip = RTA_DATA(htb[MDBA_MDB_EATTR_EHT_HOST_IP_ADDR]);
+		ip_addr = inet_ntop(af, ip, ip_buf, sizeof(ip_buf));
+		if (ip_addr) {
+			print_color_string(PRINT_ANY, ifa_family_color(af),
+					"host_ip", "    Host IP: %s", ip_addr);
+		}
+	}
+
+	/* Parse and print host MAC address */
+	if (htb[MDBA_MDB_EATTR_EHT_HOST_MAC]) {
+		const unsigned char *mac = RTA_DATA(htb[MDBA_MDB_EATTR_EHT_HOST_MAC]);
+		snprintf(mac_buf, sizeof(mac_buf),
+				"%02x:%02x:%02x:%02x:%02x:%02x",
+				mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+		mac_addr = mac_buf;
+		print_string(PRINT_ANY, "host_mac", " MAC: %s", mac_addr);
+	}
+
+	/* Parse and print filter mode */
+	if (htb[MDBA_MDB_EATTR_EHT_HOST_MODE]) {
+		mode = rta_getattr_u8(htb[MDBA_MDB_EATTR_EHT_HOST_MODE]);
+		mode_str = (mode == MCAST_INCLUDE) ? "INCLUDE" : "EXCLUDE";
+		print_string(PRINT_ANY, "mode", " Mode: %s", mode_str);
+	}
+
+	/* Parse and print number of sources */
+	if (htb[MDBA_MDB_EATTR_EHT_HOST_NSRCS]) {
+		nsrcs = rta_getattr_u32(htb[MDBA_MDB_EATTR_EHT_HOST_NSRCS]);
+		print_uint(PRINT_ANY, "nsrcs", " NSrcs: %u", nsrcs);
+	}
+
+	print_nl();
+
+	/* Print source list if present */
+	if (htb[MDBA_MDB_EATTR_EHT_HOST_SRCS]) {
+		print_eht_host_srcs(htb[MDBA_MDB_EATTR_EHT_HOST_SRCS], af);
+	}
+
+	close_json_object();
+}
+
+static void print_eht_hosts(struct rtattr *eht_attr, int af)
+{
+	struct rtattr *i;
+	int rem;
+
+	if (!eht_attr)
+		return;
+
+	open_json_array(PRINT_JSON, "eht_hosts");
+	if (!is_json_context())
+		print_string(PRINT_FP, NULL, "%s", "\nEHT Hosts:\n");
+
+	rem = RTA_PAYLOAD(eht_attr);
+	for (i = RTA_DATA(eht_attr); RTA_OK(i, rem); i = RTA_NEXT(i, rem)) {
+		print_eht_host_entry(i, af);
+	}
+
+	close_json_array(PRINT_JSON, NULL);
+}
+
 static void print_dst(const struct rtattr *dst_attr)
 {
 	SPRINT_BUF(abuf);
@@ -291,6 +412,10 @@ static void print_mdb_entry(FILE *f, int
 			     format_timer(timer, 1));
 	}
 
+	if (show_eht_hosts && tb && tb[MDBA_MDB_EATTR_EHT_HOSTS]) {
+		print_eht_hosts(tb[MDBA_MDB_EATTR_EHT_HOSTS], af);
+	}
+
 	print_nl();
 	close_json_object();
 }
@@ -442,6 +567,8 @@ int print_mdb_mon(struct nlmsghdr *n, vo
 static int mdb_show(int argc, char **argv)
 {
 	char *filter_dev = NULL;
+	bool send_dump_flags = false;
+	__u32 dump_flags = 0;
 
 	while (argc > 0) {
 		if (strcmp(*argv, "dev") == 0) {
@@ -467,8 +594,20 @@ static int mdb_show(int argc, char **arg
 	new_json_obj(json);
 	open_json_object(NULL);
 
+	/* Check if show_eht_hosts was set globally (e.g., bridge -e mdb show) */
+	if (show_eht_hosts) {
+		dump_flags |= MDBA_MDB_DUMPF_EHT;
+		send_dump_flags = true;
+	}
+
 	/* get mdb entries */
-	if (rtnl_mdbdump_req(&rth, PF_BRIDGE) < 0) {
+	if (send_dump_flags) {
+		if (rtnl_mdbdump_req_filter(&rth, PF_BRIDGE, dump_flags) < 0) {
+			perror("Cannot send dump request");
+			delete_json_obj();
+			return -1;
+		}
+	} else if (rtnl_mdbdump_req(&rth, PF_BRIDGE) < 0) {
 		perror("Cannot send dump request");
 		delete_json_obj();
 		return -1;
Index: iproute2-6.11.0/include/uapi/linux/if_bridge.h
===================================================================
--- iproute2-6.11.0.orig/include/uapi/linux/if_bridge.h
+++ iproute2-6.11.0/include/uapi/linux/if_bridge.h
@@ -608,10 +608,13 @@ enum {
 	MDBA_UNSPEC,
 	MDBA_MDB,
 	MDBA_ROUTER,
+	MDBA_MDB_EHT_DUMP,
 	__MDBA_MAX,
 };
 #define MDBA_MAX (__MDBA_MAX - 1)
 
+#define MDBA_MDB_DUMPF_EHT	(1 << 0) /* Include EHT host tracking data */
+
 enum {
 	MDBA_MDB_UNSPEC,
 	MDBA_MDB_ENTRY,
@@ -639,10 +642,34 @@ enum {
 	MDBA_MDB_EATTR_VNI,
 	MDBA_MDB_EATTR_IFINDEX,
 	MDBA_MDB_EATTR_SRC_VNI,
+	MDBA_MDB_EATTR_EHT_HOSTS,
 	__MDBA_MDB_EATTR_MAX
 };
 #define MDBA_MDB_EATTR_MAX (__MDBA_MDB_EATTR_MAX - 1)
 
+/* per mdb entry EHT host attributes */
+enum {
+	MDBA_MDB_EATTR_EHT_HOST_UNSPEC,
+	MDBA_MDB_EATTR_EHT_HOST_ENTRY,     /* Nested entry */
+	MDBA_MDB_EATTR_EHT_HOST_IP_ADDR,   /* Host IP address */
+	MDBA_MDB_EATTR_EHT_HOST_MODE,      /* INCLUDE/EXCLUDE mode */
+	MDBA_MDB_EATTR_EHT_HOST_NSRCS,     /* Number of sources */
+	MDBA_MDB_EATTR_EHT_HOST_SRCS,      /* List of source addresses */
+	MDBA_MDB_EATTR_EHT_HOST_MAC,       /* Host MAC address */
+	__MDBA_MDB_EATTR_EHT_HOST_MAX
+};
+#define MDBA_MDB_EATTR_EHT_HOST_MAX (__MDBA_MDB_EATTR_EHT_HOST_MAX - 1)
+
+/* per EHT host source address attributes */
+enum {
+	MDBA_MDB_EATTR_EHT_HOST_SRC_UNSPEC,
+	MDBA_MDB_EATTR_EHT_HOST_SRC_ENTRY,   /* Nested source entry */
+	MDBA_MDB_EATTR_EHT_HOST_SRC_ADDR,    /* Source IP address */
+	MDBA_MDB_EATTR_EHT_HOST_SRC_TIMER,   /* Source timer value */
+	__MDBA_MDB_EATTR_EHT_HOST_SRC_MAX
+};
+#define MDBA_MDB_EATTR_EHT_HOST_SRC_MAX (__MDBA_MDB_EATTR_EHT_HOST_SRC_MAX - 1)
+
 /* per mdb entry source */
 enum {
 	MDBA_MDB_SRCLIST_UNSPEC,
Index: iproute2-6.11.0/include/libnetlink.h
===================================================================
--- iproute2-6.11.0.orig/include/libnetlink.h
+++ iproute2-6.11.0/include/libnetlink.h
@@ -116,6 +116,8 @@ int rtnl_nexthop_bucket_dump_req(struct
 int rtnl_tunneldump_req(struct rtnl_handle *rth, int family, int ifindex,
 			__u8 flags)
 	__attribute__((warn_unused_result));
+int rtnl_mdbdump_req_filter(struct rtnl_handle *rth, int family, __u32 dump_flags)
+	__attribute__((warn_unused_result));
 
 struct rtnl_ctrl_data {
 	int	nsid;
Index: iproute2-6.11.0/lib/libnetlink.c
===================================================================
--- iproute2-6.11.0.orig/lib/libnetlink.c
+++ iproute2-6.11.0/lib/libnetlink.c
@@ -1652,3 +1652,24 @@ int rtnl_tunneldump_req(struct rtnl_hand
 
 	return send(rth->fd, &req, sizeof(req), 0);
 }
+
+int rtnl_mdbdump_req_filter(struct rtnl_handle *rth, int family, __u32 dump_flags)
+{
+	struct {
+		struct nlmsghdr nlh;
+		struct br_port_msg bpm;
+		char buf[1024];
+	} req = {
+		.nlh.nlmsg_len = NLMSG_LENGTH(sizeof(struct br_port_msg)),
+		.nlh.nlmsg_type = RTM_GETMDB,
+		.nlh.nlmsg_flags = NLM_F_DUMP | NLM_F_REQUEST,
+		.nlh.nlmsg_seq = rth->dump = ++rth->seq,
+		.bpm.family = family,
+	};
+
+	if (dump_flags) {
+		addattr32(&req.nlh, sizeof(req), MDBA_MDB_EHT_DUMP, dump_flags);
+	}
+
+	return send(rth->fd, &req, req.nlh.nlmsg_len, 0);
+}
